### PR TITLE
Fix rev dot_product leak

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Fixes a potential memory leak in rev implementation of `dot_product` that was introducet with refactor that started using `reverse_pass_callback`.

## Tests
No new tests.

## Side Effects
None.

## Checklist

- [x] Math issue #2025 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

File: stan/math/rev/fun/dot_product.hpp [hunk 1]
```diff
@@ -38,22 +39,27 @@ inline return_type_t<T1, T2> dot_product(const T1& v1, const T2& v2) {
...
-  const auto& v1_val = to_arena_if<!is_constant<T2>::value>(value_of(v1_col));
-  const auto& v2_val = to_arena_if<!is_constant<T1>::value>(value_of(v2_col));
+  const auto& v1_val_arena
+      = to_arena_if<!is_constant<T2>::value>(value_of(v1_col));
+  const auto& v2_val_arena
+      = to_arena_if<!is_constant<T1>::value>(value_of(v2_col));
 
-  var res(new vari(dot_product(v1_val, v2_val)));
+  var res(new vari(dot_product(
+      static_select<is_constant<T2>::value>(value_of(v1_col), v1_val_arena),
```
Comment (bbbales2) [inline stan/math/rev/fun/dot_product.hpp:48]: I think we just write out all the if statements here. It's annoying duplicating stuff, but it's easier to read code:

```cpp
double res_val = 0.0;
if(is_constant<T1>::value) {
  res_val = dot_product(v1_val_arena, value_of(v2_col));
} else if(is_constant<T2>::value) {
  res_val = dot_product(value_of(v1_col), v2_val_arena);
} else {
  res_val = dot_product(v1_val_arena, v2_val_arena);
}
var res(res_val);
```

File: stan/math/rev/fun/dot_product.hpp [hunk 2]
```diff
@@ -38,22 +39,27 @@ inline return_type_t<T1, T2> dot_product(const T1& v1, const T2& v2) {
   const auto& v1_col = as_column_vector_or_scalar(v1);
   const auto& v2_col = as_column_vector_or_scalar(v2);
 
-  const auto& v1_val = to_arena_if<!is_constant<T2>::value>(value_of(v1_col));
-  const auto& v2_val = to_arena_if<!is_constant<T1>::value>(value_of(v2_col));
+  const auto& v1_val_arena
+      = to_arena_if<!is_constant<T2>::value>(value_of(v1_col));
```
Comment (bbbales2) [inline stan/math/rev/fun/dot_product.hpp:43]: Any particular reason to switch from `auto` to `const auto&` here?

Comment (bbbales2) [reply stan/math/rev/fun/dot_product.hpp:43]: Actually the type here could be `arena_t<promote_scalar_t<double, T1>>` which makes the type explicit which is nice and comforting.

**Comment by SteveBronder** [comment]
Is this waiting on review? imo the changes look good

**Comment by bbbales2** [comment]
Was waiting on the tests to pass, but yeah this looks good

**Comment by t4c1** [comment]
@serban-nicusor-toptal I have some distribution test failures here with the error `fatal error: error in backend: IO failure on output stream: No space left on device`.

**Comment by serban-nicusor-toptal** [comment]
Hey, I have encountered the error last night too. I have changed the type of EC2 machines we use but have reverted back after.
Let me know if it will still happen after this new build. Thanks!

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal seems that did not help :/ https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-2034/3/pipeline/157

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar instances were mounting the 500 GB disk we use wrong, I've fixed that and restarted this [job](https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-2026/20/pipeline/).
Please let me know if that is still happening after this, thanks!

**Comment by bbbales2** [comment]
@serban-nicusor-toptal looks like it's still not working (at least I don't think it should be failing in distribution tests)

**Comment by serban-nicusor-toptal** [comment]
That's strange, there have been no changes to the instances besides a disk mount point change, no software changes.
I'm looking into it.

**Pull request merged**